### PR TITLE
Feature : Automatic token refresh

### PR DESCRIPTION
**Pull request by vidhutsingh**
Feature : Implement automatic session token refresh if the token expires while running a long job. This will enhance the user experience. 

Details : 
SDK internally now check for HTTP code > 400 and error message "SM_http_unauthorized" while executing CRUD operations . The CRUD operations are executed after refreshing the session token. 

Added a new function to delete the session token as it was missing with SDK. 

Unit Test:
I've written a custom test code to validate this since session tokens expire in 1801 seconds. 

Signed-off-by: vidhut-kumar-singh <vidhut-kumar.singh@hpe.com>

### CODE REVIEW

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: Is there a reason to check for anything other than 401?

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: A blank line would be nice here.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: It might be unexpected for the caller to receive an error that did not come in response to their request.  An alternative is to return the original 401 error, but that doesn't seem better.

Will we advertise the automatic session management somewhere, like in the README?  That might be the place to mention this little quirk.

File: README.md
Comment (rkumpf) [inline README.md]: Remove the word "a" in the first sentence?

It is worth mentioning get/read operations in the third sentence.

File: pkg/service/volume_service_test.go
Comment (rkumpf) [inline pkg/service/volume_service_test.go]: Could we do something simple like this to avoid the long wait?

suite.groupService.client.SessionToken = "expired"

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: Before this line, add:

client.SessionToken = ""

Otherwise a changed password can still cause infinite recursion.

File: pkg/service/volume_service_test.go
Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: Yes we can do that . thanks 

`func (suite *VolumeServiceTestSuite) TestExpiredToken() {

	volume := suite.createVolume("TestExpiredToken")
	if volume != nil {
		//Set invalid session Token 9d255b36c700ec8b56e2064e67f01c45
		suite.groupService.client.SessionToken = "9d255b36c700ec8b56e2064e67f01c45"
		suite.deleteVolume("TestExpiredToken")
	}
}`

File: README.md
Comment (vidhutsingh) [reply README.md]: I will remove "a" . Please note that this is the initial README. It will be rewritten by Michael.

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: Just want to avoid any NOS version issue where it doesn't return 401 for unauthorized error . Ideally we should use 401 .

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: Done .

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: May contain in the documentation.

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: done

File: README.md
Comment (rgcostea) [inline README.md]: This is *the* Go Software...

File: README.md
Comment (rgcostea) [inline README.md]: The SDK provides Go modules...

File: pkg/client/group_mgmt_client.go
Comment (rgcostea) [reply pkg/client/group_mgmt_client.go]: We should only do this for a 401.  All other 400 errors need to make it back to the client for assessment.

400 (bad request) - could be a malformed URL or post body
403 (forbidden) - permission issue or inability to perform an operation on some object
404 (not found) - bad url path
409 (conflict) - object in use
...

Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: @rgcostea , we are returning all the error code to the client. Here we are checking both code >= 400 and SM_http_unauthorized before considering this as an expired token scenario. So if you are certain that we don't get any other error code with any previous NOS version for unauthorized error, I will make this change.

File: pkg/client/group_mgmt_client.go
Comment (rgcostea) [inline pkg/client/group_mgmt_client.go]: goto statements are rarely a good option when you can opt to do something else.

Why is this code copied in every function (Get, Post, Put, etc.)?  Why not simply move this code into a function (e.g. verifyAuth) which checks for the 401 code and attempts to recreate the session?

Regarding the goto, please investigate the RetryCondition option in go-resty.  If that doesn't pan out, I still prefer we find another way to perform the retry rather than a goto.

File: pkg/service/group_service_factory.go
Comment (rgcostea) [inline pkg/service/group_service_factory.go]: This doesn't seem correct to me.  I prefer you just delete the token through the tokenService.  Alternatively, you can rename this function to logout I suppose.

File: pkg/client/group_mgmt_client.go
Comment (rgcostea) [reply pkg/client/group_mgmt_client.go]: I'm certain.  It's only ever a 401.  Let's keep it clean.

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: >> Why is this code copied in every function (Get, Post, Put, etc.)? Why not simply move this code into a function (e.g. verifyAuth) which checks for the 401 code and attempts to recreate the session?

We need refresh the token in case of 401 and also we need to retry the failed operation so it has to go in (Get, Post, Put etc ). 

I will check go-resty , other way could be to use for{} loop instead of goto.

**Review by suneeth51 — DISMISSED** [review]
@vidhutsingh  As @rgcostea  mentioned, definitely we can further optimize the token refresh logic and also refactor the code across all those CRUD operations and have **single common function** that first attempts perform REST request to the Array and if it encounters **401** error, then only you need to perform token refreshing.

Please refer to this CSP implementation for reference:
https://github.com/hpe-storage/common-host-libs/blob/master/storageprovider/csp/container_storage_provider.go
*func (provider *ContainerStorageProvider) invoke(request *connectivity.Request) (status int, err error)*

**Comment by vidhutsingh** [comment]
Thanks @suneeth51

**Comment by vidhutsingh** [comment]
> https://github.com/hpe-storage/common-host-libs/blob/master/storageprovider/csp/container_storage_provider.go

We shouldn't compare this with CSP implementation as here we are using resty package and each CRUD operations GET, PUT etc are implemented as different function unlike in CSP where you build the path and just invoke "provider.Client.DoJSON()"
Here is code for reference "
`response, err := client.Client.R().
		SetHeader("X-Auth-Token", client.SessionToken).
		SetBody(&DataWrapper{
			Data: payload,
		}).
		Post(url)`

I will still see how much refactoring is possible although sure we can remove GOTO and use recursion call instead.

**Comment by rgcostea** [comment]
I feel the RetryCondition is a better option.  You can perform a login, update the header token, and return true.  This will force the resty client to retry.

File: pkg/service/group_service_factory.go
Comment (vidhutsingh) [reply pkg/service/group_service_factory.go]: Ok will change to LogoutService

File: README.md
Comment (vidhutsingh) [reply README.md]: ok thanks

File: README.md
Comment (vidhutsingh) [reply README.md]: sure

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: I've lost track of what this condition is for.  I'm concerned that if the login fails at line 96, we cannot recover.  Is there any update on using RetryConditions?

**Comment by vidhutsingh** [comment]
> I feel the RetryCondition is a better option. You can perform a login, update the header token, and return true. This will force the resty client to retry.

Yes @rgcostea , resty RetryCondition is a superior option although it took me a while to understand how does it work, it is totally worth it. 
thanks . 👍

**Comment by vidhutsingh** [comment]
> > I feel the RetryCondition is a better option. You can perform a login, update the header token, and return true. This will force the resty client to retry.
> 
> Yes @rgcostea , resty RetryCondition is a superior option although it took me a while to understand how does it work, it is totally worth it.
> thanks . 👍

While doing unit testing discover an issue, the RetryCondition is not working in our case where Auth Token is refreshed. It always pic the expired token. Although workflow retry does happen possibly a bug or by design with rest since 
the request doesn't take a pointer to SessionToken it takes string value. 

```
response, err := client.Client.R().
		SetHeader("X-Auth-Token", client.SessionToken).
		SetBody(&DataWrapper{
			Data: payload,
		}).
```

**Comment by rgcostea** [comment]
What does your code look like in the RetryCondition?

**Comment by vidhutsingh** [comment]
```
Here is the code with only changes 
``// NewClient instantiates a new client to communicate with the Nimble group
func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) (*GroupMgmtClient, error) {
	// Create GroupMgmt Client
	client := resty.New()
	client.SetTLSClientConfig(&tls.Config{
		InsecureSkipVerify: true,
	})
	groupMgmtClient := &GroupMgmtClient{
		URL:       fmt.Sprintf("https://%s:5392/%s", ipAddress, apiVersion),
		Client:    client,
		WaitOnJob: waitOnJobs,
		Username:  username,
		Password:  password,
	}

	// Get session token
	sessionToken, err := groupMgmtClient.login(username, password)
	if err != nil {
		return nil, err
	}
	groupMgmtClient.SessionToken = sessionToken

	// Add retryCondition
	groupMgmtClient.Client.
		SetRetryCount(maxOpsRetries).
		AddRetryCondition(
			func(resp *resty.Response, err error) bool {
				if err == nil && resp.StatusCode() == 401 {

					err := groupMgmtClient.refreshSessionToken()
					if err != nil {
						return false
					}
					fmt.Println("RETRY WORKFLOW with new token", groupMgmtClient.SessionToken)
					return true
				}
				return false
			},
		)
	return groupMgmtClient, nil
}

// EnableDebug : enables debug logging of client request/response
func (client *GroupMgmtClient) EnableDebug() {
	client.Client.SetDebug(true)
}

// refreshSessionToken : refresh session token
func (client *GroupMgmtClient) refreshSessionToken() error {
	nClient := *client
	newClient := &nClient
	rclient := resty.New()
	rclient.SetTLSClientConfig(&tls.Config{
		InsecureSkipVerify: true,
	})
	newClient.Client = rclient
	newSessionToken, err := newClient.login(client.Username, client.Password)
	if err != nil {
		return err
	}
	// set new auth token
	fmt.Println("new session", newSessionToken)
	client.SessionToken = newSessionToken
	return nil

}``
```

**Comment by rgcostea** [comment]
It looks like you are setting the retrycondition on the Client and not on the request:

groupMgmtClient.Client vs client.Client.R() 

client.Client.R() is where the auth-token is set.

**Comment by vidhutsingh** [comment]
As per the below example code it has to be set on client 
```
retryCount := 5
	retryIntervals := make([]uint64, retryCount+1)

	// Set retry wait times that do not intersect with default ones
	retryWaitTime := time.Duration(3) * time.Second
	retryMaxWaitTime := time.Duration(9) * time.Second

	c := dc().
		SetRetryCount(retryCount).
		SetRetryWaitTime(retryWaitTime).
		SetRetryMaxWaitTime(retryMaxWaitTime).
		AddRetryCondition(
			func(r *Response, _ error) bool {
				timeSlept, _ := strconv.ParseUint(string(r.Body()), 10, 64)
				retryIntervals[attempt] = timeSlept
				attempt++
				return true
			},
		)
	_, _ = c.R().Get(ts.URL + "/set-retrywaittime-test")
```

Reference : https://github.com/go-resty/resty/blob/master/resty_test.go

**Comment by rgcostea** [comment]
Ok.  But the Response in the RetryCondition has a reference to the request.  Can't you call 

`response.request.SetHeader("X-Auth-Token", newSession)`

**Comment by vidhutsingh** [comment]
It worked.  thanks. 
How did you find out that response can also be used to set request header? that was really nice, seems quirky a bit :) 

Please check the latest code

**Comment by rgcostea** [comment]
I just read the docs: https://godoc.org/github.com/go-resty/resty#Response

**Comment by vidhutsingh** [comment]
Thanks @rgcostea . I just realised that we are also storing the rest client pointer in the groupMgtClient, we could have used it to set the header as well. Either way is fine.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: We should clearly document this choice.  It's probably not a first release requirement to do something better, but at some point we should.  I hope George's request to change the constructor to use the functional options pattern will be implemented, since it will let us make these kinds of changes without breaking API compatibility.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: Let's add a comment explaining why we create a new client.  I guess it is because the resty client gets upset if it has two requests going in parallel.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: This isn't true anymore, right?  I think the retry would already have happened at line 158.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go]: It's a shame we have to unmarshal the response twice.  Could we reasonably avoid it?

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go:320]: In several places I see logic like this:

        if !response.IsSuccess() {
                return nil, processError(response.StatusCode(), response.Body())
        }
        return processResponse(client, response, path, respHolder)

It seems redundant with this check.

File: pkg/service/volume_service_test.go
Comment (rkumpf) [inline pkg/service/volume_service_test.go]: Is this change desired?

File: pkg/service/volume_service_test.go
Comment (rkumpf) [inline pkg/service/volume_service_test.go]: What a bummer to put these checks everywhere.  Is there no better option?

**Review by rkumpf — COMMENTED** [review]
I like the new RetryCondition.  :)

File: pkg/client/group_mgmt_client.go
Comment (rgcostea) [reply pkg/client/group_mgmt_client.go]: Why do we create a new client?  I don't understand the need.

File: pkg/client/group_mgmt_client.go
Comment (rgcostea) [reply pkg/client/group_mgmt_client.go]: +1

I think the tls.Config should be an option to the GroupMgmtClientConfig.  Let the user decide if they want to skip trust to the array.

File: pkg/client/group_mgmt_client.go
Comment (rgcostea) [inline pkg/client/group_mgmt_client.go]: Delete or Post?

File: pkg/service/volume_service_test.go
Comment (rgcostea) [reply pkg/service/volume_service_test.go]: Agreed.  Why were they added?

It seems since you check for this SetupTest() above, it will fail there and none of these tests will run.

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: Yes it is not needed now since we are setting the header . Infact whole refreshSessionToken fn is not required .

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: Leftover comments , will clean up

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: Ok will fix it .

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go:320]: Oh we dont need it , let me clean up

File: pkg/service/volume_service_test.go
Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: No

File: pkg/service/volume_service_test.go
Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: It runs and panic in all the test cases . Thus had to put nil check everywhere.

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: The comment is not needed here .

File: pkg/service/volume_service_test.go
Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: No it doesnt , it runs all the test even setup failed.

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: done

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: yes

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: not needed . will cleanup .

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: We will track it via functional option pattern

File: pkg/service/volume_service_test.go
Comment (rkumpf) [reply pkg/service/volume_service_test.go]: If setup fails, with this change does the test case pass or fail?

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go:335]: I guess the 'totalRows != nil' check doesn't really serve a purpose.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go:336]: This comment isn't related to this code review, but it is a convenient place for me to mention it anyway.  :)

The pagination.go file is in a param/pagination package.  Is there a reason not to put it in the param package directly?

File: pkg/service/volume_service_test.go
Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: setup is not a test , all the test starts with Test* methods so eventually it will come to all the Test* methods even after setup has error .

Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: Let me check the documentation of suite package . will see if anything can be done about it .

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go:330]: Nit: It might be more readable if we remove the 'else' and unindent this 'return'.

File: pkg/service/volume_service_test.go
Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: Found it , we just need to exit the suite. However there is no way via suite package . pls check last commit

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go:333]: The status code is presumed to be 200 OK when we get here, right?  I suggest checking for that and returning some kind of "unexpected status code" error otherwise.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [inline pkg/client/group_mgmt_client.go:356]: We need a delimiter between the messages.

**Review by rkumpf — COMMENTED** [review]
Please consider addressing these mostly minor and/or unrelated comments in the next patch.

File: pkg/service/volume_service_test.go
Comment (rkumpf) [inline pkg/service/volume_service_test.go]: Do not include a password.  One option is to require an environment variable to be set when the test is run.

**Review by rkumpf — APPROVED** [review]

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go:336]: We need to initialze the totalRow which is not visible outside this pkg

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go:333]: Not require since error is already captured in err

File: pkg/service/volume_service_test.go
Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: That is a test code . that should be okay .

File: pkg/client/group_mgmt_client.go
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go:335]: It does since if there is error it will be nil .

**Comment by vidhutsingh** [comment]
I think all the major changes have been accommodated with the last commit.

File: pkg/service/volume_service_test.go
Comment (rkumpf) [reply pkg/service/volume_service_test.go]: It is not okay.  Email me if you would like to discuss it.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [reply pkg/client/group_mgmt_client.go:333]: I'm referring to the handling of status codes like 201 Created, 204 No Content, etc.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [reply pkg/client/group_mgmt_client.go:336]: I probably wasn't clear.  I'm suggesting moving param/pagination/pagination.go to param/pagination.go, unless there is a good reason to have a separate package for it.

File: pkg/client/group_mgmt_client.go
Comment (rkumpf) [reply pkg/client/group_mgmt_client.go:335]: I'm saying this would be enough:

		if params != nil && params.Page != nil {
			params.Page.TotalRows = totalRows

If totalRows is nil, I would think params.Page.TotalRows should be nil also.

Comment (rkumpf) [reply pkg/client/group_mgmt_client.go:335]: I'm reconsidering this one.  I think what you mean is when an error occurs, we should preserve the prior TotalRows value even though we'll be returning something that didn't come back in the API response.  The client program will see that an error is returned and know not to inspect the TotalRows value.  That seems reasonable to me.  We can close this issue.

File: pkg/service/volume_service_test.go
Comment (vidhutsingh) [reply pkg/service/volume_service_test.go]: removed it

File: pkg/service/volume_service_test.go
Comment (rkumpf) [reply pkg/service/volume_service_test.go]: It's back! ;)

**Review by rkumpf — APPROVED** [review]

**Review by rkumpf — APPROVED** [review]

**Comment by vidhutsingh** [comment]
@rkumpf, the merge is still not approved.

**Comment by vidhutsingh** [comment]
@rgcostea , there is one approver but still, the merge is not allowed. is there any setting change?

**Comment by rkumpf** [comment]
The merge is blocked because Suneeth has an outstanding request for changes.  His comment seems to be addressed acceptably in the updated patch, but I don't perceive sufficient urgency to warrant the bold step of dismissing his review.

**Comment by vidhutsingh** [comment]
> The merge is blocked because Suneeth has an outstanding request for changes. His comment seems to be addressed acceptably in the updated patch, but I don't perceive sufficient urgency to warrant the bold step of dismissing his review.

Nobody suggested that . It just that I wasn't aware of what is blocking the merge. now it's clear. ;)

**Comment by rkumpf** [comment]
I know you didn't suggest it, but I was considering it nonetheless....

**Comment by vidhutsingh** [comment]
@suneeth51 , the merge is blocked because of your review change request. If you don't have any further comments, please approve. 

Very poor feature of GitHub as any reviewer can block the merge. Bummer!

**Comment by vidhutsingh** [comment]
I am blocked due to this .

**Comment by vidhutsingh** [comment]
@rkumpf , thank you!

**Comment by suneeth51** [comment]
@rkumpf Thanks for taking care of this to unblock @vidhutsingh. The changes looks good.

**Pull request merged**
